torch_geometric/utils/negative_sampling.py

```diff
@@ -99,23 +104,34 @@ def structured_negative_sampling(edge_index, num_nodes=None):
         edge_index (LongTensor): The edge indices.
         num_nodes (int, optional): The number of nodes, *i.e.*
             :obj:`max_val + 1` of :attr:`edge_index`. (default: :obj:`None`)
+        contains_neg_self_loops (bool, optional): If set to
+            :obj:`False`, sampled negative edges will not contain self loops.
+            It is independent of :attr:`edge_index`. (default: :obj:`True`)
 
     :rtype: (LongTensor, LongTensor, LongTensor)
     """
     num_nodes = maybe_num_nodes(edge_index, num_nodes)
 
+    assert is_neg_samp_feasible(edge_index, 'structure', num_nodes, False,
+                                contains_neg_self_loops)
+
+    self_loops = torch.arange(num_nodes) * (num_nodes + 1)
     i, j = edge_index.to('cpu')
     idx_1 = i * num_nodes + j
```

Comment:
what do you think of modifying `idx_1` if `contains_neg_self_loops` is `false`. That way the if condition in the while loop will not be needed.
```suggestion
    self_loops = torch.arange(num_nodes) * (num_nodes + 1)
    i, j = edge_index.to('cpu')
    idx_1 = i * num_nodes + j
    idx_1 = idx_1 if contains_neg_self_loops else torch.cat([idx_1, self_loops])
```